### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by AlexejheroYTB**
<h2><i>The refactorer strikes again!!</i></h2>

### Changes made in this pull request

- Overall cleanup in the project structure to make it less confusing for newer contributors and make it easier to add new features without throwing them in the project in random places. 
    - Classes are now grouped by the main feature they represent, followed by sub-feature or class type.
        - For example, `Networking/Rpcs` contains all of the rpc-related classes
        - `GUI/Extensions` contains extension methods for Unity UI, like Rect or Texture-related stuff
    - Cleaned up the "extension" files which, even though they were supposed to be grouped by category, seemed to have been scrambled by someone
    - Cleaned up the "patches" files which also contained extensions and random monobehaviours
- Updated BepInEx and GameLibs dependencies

### CODE REVIEW

**Review by simonkellly — APPROVED** [review]
Very good changes

**Comment by edqx** [comment]
Very good changes
AlexejheroYTB reacted with thumbs up emoji
natsu-k reacted with

File: Reactor/Networking/Rpc/CustomRpcManager.cs [hunk 1]
```diff
@@ -17,8 +23,8 @@ public UnsafeCustomRpc Register(UnsafeCustomRpc customRpc)
     {
         customRpc.Manager = this;
         _list.Add(customRpc);
-        _map.GetOrCreate(customRpc.InnerNetObjectType, _ => new Dictionary<Mod, Dictionary<uint, UnsafeCustomRpc>>())
-            .GetOrCreate(customRpc.Mod, _ => new Dictionary<uint, UnsafeCustomRpc>())
+        _map.GetOrCreate(customRpc.InnerNetObjectType, static _ => new Dictionary<Mod, Dictionary<uint, UnsafeCustomRpc>>())
+            .GetOrCreate(customRpc.Mod, static _ => new Dictionary<uint, UnsafeCustomRpc>())
```
Comment (AlexejheroYTB) [inline Reactor/Networking/Rpc/CustomRpcManager.cs:27]: Wtf is this static arrow functions

File: Reactor/Networking/Rpc/CustomRpcManager.cs [hunk 2]
```diff
@@ -28,4 +34,40 @@ public UnsafeCustomRpc Register(UnsafeCustomRpc customRpc)
...
+        public static IEnumerable<MethodBase> TargetMethods()
+        {
+            return InnerNetObjectTypes.Select(x => x.GetMethod(nameof(InnerNetObject.HandleRpc), AccessTools.allDeclared)).Where(m => m != null)!;
+        }
+
+        public static bool Prefix(InnerNetObject __instance, [HarmonyArgument(0)] byte callId, [HarmonyArgument(1)] MessageReader reader)
+        {
+            return !HandleRpc(__instance, callId, reader);
+        }
+    }
```
Comment (AlexejheroYTB) [inline Reactor/Networking/Rpc/CustomRpcManager.cs:72]: Moving the patch back in here?

File: Reactor/Patches/Fixes/CursorPosPatch.cs [hunk 3]
```diff
@@ -8,12 +8,10 @@ namespace Reactor.Patches.Fixes;
...
 /// "Fixes" an issue where empty TextBoxes have wrong cursor positions
 /// </summary>
-[HarmonyPatch]
+[HarmonyPatch(typeof(TextMeshProExtensions), nameof(TextMeshProExtensions.CursorPos))]
 internal static class CursorPosPatch
 {
-    [HarmonyPatch(typeof(TextMeshProExtensions), nameof(TextMeshProExtensions.CursorPos))]
-    [HarmonyPrefix]
-    public static bool FixCursorPosPatch(TextMeshPro self, ref Vector2 __result)
+    public static bool Prefix(TextMeshPro self, ref Vector2 __result)
```
Comment (AlexejheroYTB) [inline Reactor/Patches/Fixes/CursorPosPatch.cs:14]: By changing this you end up with inconsistencies with other patches

File: Reactor/Utilities/RegionInfoWatcher.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        Watcher.EnableRaisingEvents = true;
+    }
+
+    public void Dispose()
+    {
+        Watcher.Dispose();
+    }
+
+    [HarmonyPatch]
+    internal static class WriteAllTextPatch
```
Comment (AlexejheroYTB) [inline Reactor/Utilities/RegionInfoWatcher.cs:59]: Again, moving patches into other files

File: Reactor/Patches/QOL/BetterGuestPatch.cs [hunk 5]
```diff
@@ -1,16 +0,0 @@
...
-internal static class BetterGuestPatch
-{
-    [HarmonyPatch(typeof(SaveManager), nameof(SaveManager.ChatModeType), MethodType.Getter)]
-    [HarmonyPrefix]
-    public static bool ChatModeTypePatch(out QuickChatModes __result)
-    {
-        __result = QuickChatModes.FreeChatOrQuickChat;
-        return false;
-    }
-}
```
Comment (AlexejheroYTB) [inline Reactor/Patches/QOL/BetterGuestPatch.cs]: What happened to this?

File: Reactor/Networking/Rpc/CustomRpcManager.cs (hunk 2 above)
Comment (js6pak) [reply Reactor/Networking/Rpc/CustomRpcManager.cs:72]: It's related to RPCs so it fits in `Reactor.Networking.Rpc` more than `Reactor.Networking.Patches`, and I don't feel like it's big enough to warrant a new file. Also `CustomRpcManager.HandleRpc` is public on purpose, you can use it in custom net objects if you want custom rpcs on them.

File: Reactor/Utilities/RegionInfoWatcher.cs (hunk 4 above)
Comment (js6pak) [reply Reactor/Utilities/RegionInfoWatcher.cs:59]: Again, it's small, very inherently specific to `RegionInfoWatcher` and useless outside of it.

File: Reactor/Patches/QOL/BetterGuestPatch.cs (hunk 5 above)
Comment (js6pak) [reply Reactor/Patches/QOL/BetterGuestPatch.cs]: It was quite hacky and I'd rather just move it to [Reactor.RemoveAccounts](https://github.com/NuclearPowered/Reactor.RemoveAccounts) which I'll push after this PR

**Pull request merged**